File: src/services/dictionary.ts
import request from '../api/request'
import { API_ENDPOINTS } from '../api/endpoints'
import type { ApiResponse } from '@sstcp/shared'

export interface Dictionary {
  id: number
  dict_type: string
  dict_key: string
  dict_value: string
  dict_label: string
  sort_order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface PaginatedResponse {
  code: number
  message: string
  data: {
    content: Dictionary[]
    totalElements: number
    totalPages: number
    size: number
    number: number
    first: boolean
    last: boolean
  }
}

export const dictionaryService = {
  async getList(params?: {
    page?: number
    size?: number
    dict_type?: string
  }): Promise<PaginatedResponse> {
    return await request.get(API_ENDPOINTS.DICTIONARY.LIST, { params })
  },

  async getByType(dictType: string): Promise<ApiResponse<Dictionary[]>> {
    return await request.get(API_ENDPOINTS.DICTIONARY.BY_TYPE(dictType))
  },

  async getAll(dictType?: string): Promise<ApiResponse<Dictionary[]>> {
    return await request.get('/dictionary/all/list', { params: { dict_type: dictType } })
  },

  async getById(id: number): Promise<ApiResponse<Dictionary>> {
    return await request.get(API_ENDPOINTS.DICTIONARY.DETAIL(id))
  },

  async create(data: {
    dict_type: string
    dict_key: string
    dict_value: string
    dict_label: string
    sort_order?: number
    is_active?: boolean
  }): Promise<ApiResponse<Dictionary>> {
    return await request.post(API_ENDPOINTS.DICTIONARY.LIST, data)
  },

  async update(
    id: number,
    data: {
      dict_type?: string
      dict_key?: string
      dict_value?: string
      dict_label?: string
      sort_order?: number
      is_active?: boolean
    }
  ): Promise<ApiResponse<Dictionary>> {
    return await request.put(API_ENDPOINTS.DICTIONARY.DETAIL(id), data)
  },

  async delete(id: number): Promise<ApiResponse<null>> {
    return await request.delete(API_ENDPOINTS.DICTIONARY.DETAIL(id))
  },
}

export const dictionaryTypes = {
  TEMPORARY_REPAIR_STATUS: 'temporary_repair_status',
  SPOT_WORK_STATUS: 'spot_work_status',
  PERIODIC_INSPECTION_STATUS: 'periodic_inspection_status',
  MAINTENANCE_PLAN_STATUS: 'maintenance_plan_status',
  MAINTENANCE_EXECUTION_STATUS: 'maintenance_execution_status',
  MAINTENANCE_PLAN_TYPE: 'maintenance_plan_type',
  SPARE_PARTS_STATUS: 'spare_parts_status',
} as const

File: backend-python/app/api/v1/dictionary.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, HTTPException, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.services.dictionary import DictionaryService
from app.schemas.common import ApiResponse, PaginatedResponse


router = APIRouter(prefix="/dictionary", tags=["Dictionary Management"])


@router.get("/type/{dict_type}", response_model=ApiResponse)
def get_dictionary_by_type(
    dict_type: str,
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    items = service.get_by_type(dict_type)
    return ApiResponse(
        code=200,
        message="success",
        data=[item.to_dict() for item in items]
    )


@router.get("/all/list", response_model=ApiResponse)
def get_all_dictionaries_unpaginated(
    dict_type: Optional[str] = Query(None, description="Dictionary type filter"),
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    items = service.get_all_unpaginated(dict_type)
    return ApiResponse(
        code=200,
        message="success",
        data=[item.to_dict() for item in items]
    )


@router.get("", response_model=ApiResponse)
def get_dictionaries_list(
    page: int = Query(0, ge=0, description="Page number, starts from 0"),
    size: int = Query(10, ge=1, le=2000, description="Page size"),
    dict_type: Optional[str] = Query(None, description="Dictionary type filter"),
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    items, total = service.get_all(page=page, size=size, dict_type=dict_type)
    items_dict = [item.to_dict() for item in items]
    return ApiResponse(
        code=200,
        message="success",
        data={
            'content': items_dict,
            'totalElements': total,
            'totalPages': (total + size - 1) // size,
            'size': size,
            'number': page,
            'first': page == 0,
            'last': page >= (total + size - 1) // size
        }
    )


@router.get("/{id}", response_model=ApiResponse)
def get_dictionary_by_id(
    id: int,
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    dictionary = service.get_by_id(id)
    return ApiResponse(
        code=200,
        message="success",
        data=dictionary.to_dict()
    )


@router.post("", response_model=ApiResponse, status_code=status.HTTP_201_CREATED)
def create_dictionary(
    dto: dict,
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    dictionary = service.create(dto)
    return ApiResponse(
        code=200,
        message="Created successfully",
        data=dictionary.to_dict()
    )


@router.put("/{id}", response_model=ApiResponse)
def update_dictionary(
    id: int,
    dto: dict,
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    dictionary = service.update(id, dto)
    return ApiResponse(
        code=200,
        message="Updated successfully",
        data=dictionary.to_dict()
    )


@router.delete("/{id}", response_model=ApiResponse)
def delete_dictionary(
    id: int,
    db: Session = Depends(get_db)
):
    service = DictionaryService(db)
    service.delete(id)
    return ApiResponse(
        code=200,
        message="Deleted successfully",
        data=None
    )
